src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -324,42 +324,76 @@ impl KvmVcpu {
     ///
-    /// * `msr_index_list`: List of MSR indices.
+    /// * `msr_index_iter`: Iterator over MSR indices.
     ///
     /// # Errors
     ///
     /// * When [`kvm_bindings::Msrs::new`] returns errors.
     /// * When [`kvm_ioctls::VcpuFd::get_msrs`] returns errors.
     /// * When the return value of [`kvm_ioctls::VcpuFd::get_msrs`] (the number of entries that
     ///   could be gotten) is less than expected.
-    fn get_msr_chunks(&self, msr_index_list: &[u32]) -> Result<Vec<Msrs>, KvmVcpuError> {
-        let mut msr_chunks: Vec<Msrs> = Vec::new();
-
-        for msr_index_chunk in msr_index_list.chunks(KVM_MAX_MSR_ENTRIES) {
-            let mut msrs = Msrs::new(msr_index_chunk.len())?;
-            let msr_entries = msrs.as_mut_slice();
-            assert_eq!(msr_index_chunk.len(), msr_entries.len());
-            for (pos, index) in msr_index_chunk.iter().enumerate() {
-                msr_entries[pos].index = *index;
-            }
-
-            let expected_nmsrs = msrs.as_slice().len();
-            let nmsrs = self
-                .fd
-                .get_msrs(&mut msrs)
-                .map_err(KvmVcpuError::VcpuGetMsrs)?;
-            if nmsrs != expected_nmsrs {
-                return Err(KvmVcpuError::VcpuGetMsr(msr_index_chunk[nmsrs]));
-            }
-
-            msr_chunks.push(msrs);
+    fn get_msr_chunks(
+        &self,
+        mut msr_index_iter: impl ExactSizeIterator<Item = u32>,
+    ) -> Result<Vec<Msrs>, KvmVcpuError> {
+        let num_chunks = msr_index_iter.len() / KVM_MAX_MSR_ENTRIES;
+        let excess = msr_index_iter.len() % KVM_MAX_MSR_ENTRIES;
+
+        let mut msr_chunks: Vec<Msrs> = Vec::with_capacity(num_chunks);
+
+        for _ in 0..num_chunks {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, KVM_MAX_MSR_ENTRIES)?;
+            msr_chunks.push(chunk);
+        }
+        if excess != 0 {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, excess)?;
+            msr_chunks.push(chunk);
         }
 
         Self::fix_zero_tsc_deadline_msr(&mut msr_chunks);
 
         Ok(msr_chunks)
     }
 
+    /// Get single MSR chunk for the given MSR index iterator with
+    /// specified maximum length.
+    ///
+    /// # Arguments
+    ///
+    /// * `msr_index_iter`: Iterator over MSR indices.
+    /// * `max_chunk_len`: Maximum lenght of a chunk to get.
+    ///
+    /// # Errors
+    ///
+    /// * When [`kvm_bindings::Msrs::new`] returns errors.
+    /// * When [`kvm_ioctls::VcpuFd::get_msrs`] returns errors.
+    /// * When the return value of [`kvm_ioctls::VcpuFd::get_msrs`] (the number of entries that
+    ///   could be gotten) is less than expected.
+    pub fn get_msr_chunk(
+        &self,
+        msr_index_iter: impl ExactSizeIterator<Item = u32>,
+        max_chunk_len: usize,
+    ) -> Result<Msrs, KvmVcpuError> {
+        let chunk_size = msr_index_iter.len().min(max_chunk_len);
+        let chunk_iter = msr_index_iter.take(chunk_size);
```

Comment:
why does it have to be `max_chunk_len` rather than just `chunk_len`?
Isn't the exact length of chunk calculated in `get_msr_chunks()`?

```suggestion
        chunk_len: usize,
    ) -> Result<Msrs, KvmVcpuError> {
        let chunk_iter = msr_index_iter.take(chunk_len);
```

Reply:
I was thinking about the case when iterator has less elements than the chunk size. In this case the `Msrc::new` will be given a bigger size than iterator has indexes. And I assume that KVM will not be happy to receive this half populated object. But I guess you are right, the `get_msr_chunks` should give correct size all the time. I will update this.

Reply:
indeed. you're also right. it might be better to have the note in the function's comment.